### DISCUSSION TITLE
Generic type for userData in WebSocket for Typescript

### DISCUSSION THREAD

**Discussion by yelliver** [uNetworking/uWebSockets.js#521] [Ideas]
In C++ version, uWebSockets has getUserData() while uWebSockets .js has `[key: string]: any;`
But their types are any & is not recommended in Typescript. What about this:
```
interface WebSocket<T> 
{ 
    userData: T;
}
```
We will have the consistent API between C++ & js versions as well as maintain the static type definition for Typescript.
We can also keep the `[key: string]: any;` for who want the convenience of weak typing

**Comment by SuspiciousLookingOwl** [comment]
Consider following [changes](https://github.com/uNetworking/uWebSockets.js/compare/master...SuspiciousLookingOwl:add-userdata?expand=1#diff-0d38ee7ec25681cdba88d40ada03018298549ba7c62cccade958bb4ba4e863a8).

Usage:
```ts
import { App } from "uWebSockets.js";

interface UserData {
	id: number;
	name: string;
}

const app = App<UserData>();

app.ws("/*", {
	open: (ws) => {
		ws.userData = {
			id: 0,
			name: "Owl",
		};
	},
	message: (ws) => {
		ws.userData.id; // number
		ws.userData.name; // string
		
		ws.userData.id = "Hello"; // TypeScript compilation error because `id` supposed to be `number`, not `string`
	},
});
```

---

I think this is a neat feature, temporary work around (that I use) is to create a `.d.ts` file on your project:
```ts
import "uWebSockets.js";

interface UserData {
	id: number;
	name: string;
}

declare module "uWebSockets.js" {
	interface WebSocket {
		userData: UserData;
	}
}
```
then you can use it just like the example above (without the generic type when calling `App()`)

**Comment by ghost** [comment]
It's a known problem. TS is a separate layer above JS -> whatever happens in TS, we can't know about or control. We only know about and can change V8/JS. So making template functions that take type UserData is impossible since that information only exists in TS (which is not part of V8/JS). So you can't really implement a typed interface like that.

**Reply by SuspiciousLookingOwl** [reply]
I think what @yelliver suggested is to add generic type for `ws.userData` just for better development experience when using TypeScript. Instead of dealing with `any` type when accessing `ws.userData`, developers can provide its own typing for `ws.userData` so they get that sweet auto code completion when dealing with `ws.userData`

**Reply by yelliver** [reply]
@alexhultman-2fa can you please check @SuspiciousLookingOwl  code above, it is what I want
